File: language-textmate/src/main/java/org/eclipse/tm4e/languageconfiguration/internal/model/CompleteEnterAction.java
/**
 * Copyright (c) 2015-2017 Angelo ZERR.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 * Angelo Zerr <angelo.zerr@gmail.com> - initial API and implementation
 */
package org.eclipse.tm4e.languageconfiguration.internal.model;

import org.eclipse.jdt.annotation.Nullable;
import org.eclipse.tm4e.core.internal.utils.StringUtils;

/**
 * @see <a href=
 *      "https://github.com/microsoft/vscode/blob/8e2ec5a7ee1ae5500c645c05145359f2a814611c/src/vs/editor/common/languages/languageConfiguration.ts#L250">
 *      github.com/microsoft/vscode/blob/main/src/vs/editor/common/languages/languageConfiguration.ts#L250</a>
 */
public final class CompleteEnterAction extends EnterAction {

	/**
	 * The line's indentation minus removeText
	 */
	public final String indentation;

	public CompleteEnterAction(IndentAction indentAction, @Nullable String appendText, @Nullable Integer removeText,
			final String indentation) {
		super(indentAction, appendText, removeText);
		this.indentation = indentation;
	}

	@Override
	public String toString() {
		return StringUtils.toString(this, sb -> sb
				.append("indentAction=").append(indentAction).append(", ")
				.append("appendText=").append(appendText).append(", ")
				.append("removeText=").append(removeText).append(", ")
				.append("indentation=").append(indentation));
	}
}

File: language-monarch/src/main/java/io/github/rosemoe/sora/langs/monarch/languageconfiguration/model/CompleteEnterAction.kt
/*******************************************************************************
 * Copyright (c) 2015-2017 Angelo ZERR.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 * <p>
 * SPDX-License-Identifier: EPL-2.0
 * <p>
 * Contributors:
 * Angelo Zerr <angelo.zerr@gmail.com> - initial API and implementation
 * Sebastian Thomschke (Vegard IT GmbH) - add previousLineText support
 *
 ******************************************************************************/

package io.github.rosemoe.sora.langs.monarch.languageconfiguration.model

data class CompleteEnterAction(
    /**
     * Describe what to do with the indentation.
     */
    val indentAction: Int,
    /**
     * Describes text to be appended after the new line and after the indentation.
     */
    val appendText: String,
    /**
     * Describes the number of characters to remove from the new line's indentation.
     */
    val removeText: Int?,
    /**
     * The line's indentation minus removeText
     */
    val indentation: String
)